metagraph/src/cli/assemble.cpp

```diff
@@ -17,66 +23,141 @@ using mtg::common::logger;
+    tsl::hopscotch_set<std::string> shared_foreground_labels;
+    tsl::hopscotch_set<std::string> shared_background_labels;
 
-                if (num_out_labels < config->label_mask_out_fraction
-                                        * config->label_mask_out.size())
-                    return false;
+    for (const Json::Value &group : diff_json["groups"]) {
+        if (group["shared_labels"]) {
+            shared_foreground_labels.clear();
+            shared_background_labels.clear();
+            bool has_overlap = false;
 
-                size_t num_total_labels = anno_graph.get_labels(index).size();
+            for (const Json::Value &in_label : group["shared_labels"]["in"]) {
+                if (!shared_foreground_labels.emplace(in_label.asString()).second)
+                    has_overlap = true;
+            }
 
-                return num_total_labels - num_in_labels - num_out_labels
-                            <= config->label_other_fraction * num_total_labels;
-            },
-            std::max(1u, get_num_threads())
-        )
-    );
-}
+            for (const Json::Value &out_label : group["shared_labels"]["out"]) {
+                const auto [it, inserted] = shared_background_labels.emplace(out_label.asString());
+                if (!inserted || shared_foreground_labels.count(*it))
+                    has_overlap = true;
+
+            }
+
+            if (has_overlap)
+                logger->warn("Shared label sets overlap");
+
+            check_labels(shared_foreground_labels, anno_graph);
+            check_labels(shared_background_labels, anno_graph);
+        }
+
+        if (!group["experiments"])
+            throw std::runtime_error("Missing experiments in group");
+
+        for (const Json::Value &experiment : group["experiments"]) {
+            DifferentialAssemblyConfig diff_config = parse_diff_config(
+                experiment,
+                anno_graph.get_graph().get_mode() == DeBruijnGraph::CANONICAL
+            );
+
+            foreground_labels.clear();
+            background_labels.clear();
+            bool has_overlap = false;
+
+            for (const Json::Value &in_label : experiment["in"]) {
+                const auto [it, inserted] = foreground_labels.emplace(in_label.asString());
+                if (!inserted || shared_foreground_labels.count(*it) || shared_background_labels.count(*it))
+                    has_overlap = true;
+            }
+
+            for (const Json::Value &out_label : experiment["out"]) {
+                const auto [it, inserted] = background_labels.emplace(out_label.asString());
+                if (!inserted || foreground_labels.count(*it)
+                        || shared_foreground_labels.count(*it)
+                        || shared_background_labels.count(*it)) {
+                    has_overlap = true;
+                }
+            }
 
+            if (has_overlap)
+                logger->warn("Label sets overlap");
```

Comment:
It's semantically strange to say that a given label should be both included and excluded from an experiment (and the implementation makes the assumption that these sets are disjoint).

But I guess it would be useful to print them so the user can know where to look in their config file.